### PR TITLE
Buffer based OLED panning, write byte to buffer at arbitrary index

### PR DESCRIPTION
**Pull request by brickbots**
A duet of functions to facilitate sophisticated graphical displays such as time series graphing using the OLED library.

## Description

A function to pan the contents of an OLED display left or right by manipulating the buffer. This is distinct from the scrolling features provided by the driver in that the program can control speed by panning the display column by column.

A function to write a byte at an arbitrary index of the OLED display buffer to allow for modifying discreet areas of the display.

## Types of Changes

- [x] Core
- [ ] Bugfix
- [ ] New feature
- [X] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Checklist

- [X] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [X] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [X] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by drashna** [comment]
@XScorpion2 look good to you?

File: drivers/oled/oled_driver.c [hunk 1]
```diff
@@ -431,6 +431,31 @@ void oled_write_ln(const char *data, bool invert) {
...
+                oled_buffer[i] = oled_buffer[i+1];
+            }
+        } else {
+            for (uint16_t x = OLED_DISPLAY_WIDTH -1; x > 0; x--) {
+                i = y * OLED_DISPLAY_WIDTH + x;
+                oled_buffer[i] = oled_buffer[i-1];
+            }
+        }
+    }
+    oled_dirty = 65535;
```
Comment (tzarc) [inline drivers/oled/oled_driver.c]: lets make this `~((OLED_BLOCK_TYPE)0)`, just in case someone needs to bump it to 32-bit.

Comment (brickbots) [reply drivers/oled/oled_driver.c]: Super good note, thank you @tzarc !  I've made a commit to implement this change.

File: drivers/oled/oled_driver.h [hunk 2]
```diff
@@ -200,7 +200,11 @@ void oled_write(const char *data, bool invert);
 // Advances the cursor to the next page, wiring ' ' to the remainder of the current page
 void oled_write_ln(const char *data, bool invert);
 
+// Pans the buffer to the right (or left by passing true) by moving contents of the buffer
+void oled_pan(bool left);
```
Comment (zvecr) [inline drivers/oled/oled_driver.h]: `docs/feature_oled_driver.md` should be updated with the 2 new functions

Comment (brickbots) [reply drivers/oled/oled_driver.h]: Thanks for the review and for pointing this out.  I've added this to the docs as indicated

**Comment by XScorpion2** [comment]
Not a fan of the pan function. The SSD1306 driver already supports direction (which is implemented) and speed control (not implemented). The benefit of sticking with driver provided features is that the keyboard mcu is freed up for keyboard work, not controlling the oled / sending new buffers to the oled.

**Comment by brickbots** [comment]
> Not a fan of the pan function. The SSD1306 driver already supports direction (which is implemented) and speed control (not implemented).

Hi @XScorpion2 .  Thanks for your feedback and review.  I did investigate the hardware based scrolling, but it did not seem suitable for my use case.  For reference, here is the effect I was after (see frame left display)
https://youtu.be/wHxEOVDh9UE

Here I am using the pan function to move the contents of the buffer over one column, such that I can write a new column on the left hand side and then repeat to generate this time-series graph.

The difficulties with using the hardware scrolling for this use case were:
* Inability to assure a single column scroll: This can be potentially be mitigated if we fully exposed the scroll setup and such
* Mismatch of in-memory buffer with screen contents: Once you start scrolling, the bytes in the buffer are no longer directly represented at a specific position on the display, which makes it difficult to write a byte to a specific position on the screen.  Perhaps with careful attention to the scroll parameter setup and only scrolling a set number of columns at a time the offset could be tracked in software, but any reading/writing to the buffer would then be encumbered by this offset and I suspect for sanity you'd want to actually shift the buffer as well so that it represents the display configuration.

With your experience maybe you see a way around these difficulties to support this use-case with the hardware scrolling?

It's definitely arguable that this sort of functionality could just live in user-space, and does not need to be hoisted into the oled_driver file.  I was on the fence with this and thought about just adding the raw read/write byte function here and then leave the rest in user space.  However it seemed like it might be generally useful enough and unlock some interesting functionality.  

If you feel strongly, I'm fine with removing this function and potentially writing a helper function instead to make the reading the contents of the buffer more abstract.  Then this higher-level functionality could be moved back into user space as there would be a good interface for raw reading/writing to the buffer.

Just let me know your thoughts on a way to approach this with hardware scrolling, or if you feel like this should be sent back to userspace.  Cheers!

**Comment by XScorpion2** [comment]
Holy shit that is a cool effect and ya definitely not one you can use with the hardware scrolling.
So I do take back my opinion on this. I see the reason for these two methods. I would still note in the docs to not use this function if you are just looking for basic scrolling functionality, and maybe request that you expose hardware scroll speed as well so there is a clear alternative / best path if they just need to adjust the speed.

**Comment by brickbots** [comment]
> Holy shit that is a cool effect and ya definitely not one you can use with the hardware scrolling.
Thanks!  It's probably not all the useful, but it's neat eye-candy :-) 

>  I would still note in the docs to not use this function if you are just looking for basic scrolling functionality, and maybe request that you expose hardware scroll speed as well so there is a clear alternative / best path if they just need to adjust the speed.

These are two great suggestions!  I definitely would not want people confusing the two, so I'll update the docs to help clarify.  

I'm keen to learn more about these oled displays and have a few other visualization ideas, so I'd be willing to update/expand the scroll functions to provide more control.  Once these couple of pull requests are merged, I'll open a branch and try my luck with some of the scroll setup.  Looking at the docs you linked (thank you!) there is definitely a lot of control there to be exposed.

**Comment by zvecr** [comment]
Thanks!

**Pull request merged**